unit_tests/MacroTranslTests.d

```diff
@@ -0,0 +1,475 @@
+extern (C):
+
+extern (D) auto foo(T)(in T x)
+{
+    return x.a;
+}
+
+extern (D) auto boo(T)(in T x)
+{
+    return x.a;
+}
+D");
+
+}
+
+// Translate index operator.
+unittest
+{
+    assertTranslates(q"C
+#define foo(x) x[32]
+C", q"D
+extern (C):
+
+extern (D) auto foo(T)(in T x)
+{
+    return x[32];
+}
+D");
+
+}
+
+// Translate increment decrement operators.
+unittest
+{
+    assertTranslates(q"C
+#define foo(x) ++x
+#define bar(x) --x
+#define baz(x) x++
+#define qux(x) x--
+C", q"D
+extern (C):
+
+extern (D) auto foo(T)(in T x)
+{
+    return ++x;
+}
+
+extern (D) auto bar(T)(in T x)
+{
+    return --x;
+}
+
+extern (D) auto baz(T)(in T x)
+{
+    return x++;
+}
+
+extern (D) auto qux(T)(in T x)
+{
+    return x--;
+}
+D");
+
+}
+
+// Translate type dependent macros.
+unittest
+{
+    assertTranslates(q"C
+
+typedef int uint_32;
+
+#define ROWBYTES(pixel_bits) (uint_32)(pixel_bits)
+C", q"D
+extern (C):
+
+alias int uint_32;
+
+extern (D) uint_32 ROWBYTES(T)(in T pixel_bits)
+{
+    return cast(uint_32) pixel_bits;
+}
+D");
+
+}
+
+// Disambiguate between constant and function versions of macros.
+unittest
+{
+    assertTranslates(q"C
+#define FOO 0
+#define BAR(FOO) FOO
+#define BAZ (FOO)
+C", q"D
+extern (C):
+
+enum FOO = 0;
+
+extern (D) auto BAR(T)(in T FOO)
+{
+    return FOO;
+}
+
+enum BAZ = FOO;
+D");
+
+}
+
+// Some example from libpng.
+unittest
+{
+    assertTranslates(q"C
+#define COLOR_DIST(c1, c2) (abs((int)((c1).red) - (int)((c2).red)) + \
+   abs((int)((c1).green) - (int)((c2).green)) + \
+   abs((int)((c1).blue) - (int)((c2).blue)))
+C", q"D
+extern (C):
+
+extern (D) auto COLOR_DIST(T)(in T c1, in T c2)
+{
+    return abs(cast(int) c1.red - cast(int) c2.red) + abs(cast(int) c1.green - cast(int) c2.green) + abs(cast(int) c1.blue - cast(int) c2.blue);
+}
+D");
+
+    assertTranslates(q"C
+typedef unsigned int uint_32;
+
+#define ROWBYTES(pixel_bits, width) \
+    ((pixel_bits) >= 8 ? \
+    ((width) * (((uint_32)(pixel_bits)) >> 3)) : \
+    (( ((width) * ((uint_32)(pixel_bits))) + 7) >> 3) )
+C", q"D
+extern (C):
+
+alias uint uint_32;
+
+extern (D) auto ROWBYTES(T)(in T pixel_bits, in T width)
+{
+    return pixel_bits >= 8 ? (width * ((cast(uint_32) pixel_bits) >> 3)) : (((width * (cast(uint_32) pixel_bits)) + 7) >> 3);
+}
+D");
+
+    assertTranslates(q"C
+#define OUT_OF_RANGE(value, ideal, delta) \
+        ( (value) < (ideal)-(delta) || (value) > (ideal)+(delta) )
+C", q"D
+extern (C):
+
+extern (D) auto OUT_OF_RANGE(T)(in T value, in T ideal, in T delta)
+{
+    return value < ideal - delta || value > ideal + delta;
+}
+D");
+
+    assertTranslates(q"C
+typedef unsigned uint_32;
+typedef unsigned long long foo_t;
+
+#define UINT_31_MAX ((uint_32)0x7fffffffL)
+#define UINT_32_MAX ((uint_32)(-1))
+#define FOO_MAX ((foo_t)(-1))
+#define MAX_UINT UINT_31_MAX
+C", q"D
+extern (C):
+
+alias uint uint_32;
+alias ulong foo_t;
+
+enum UINT_31_MAX = cast(uint_32) 0x7fffffffL;
+enum UINT_32_MAX = cast(uint_32) (-1);
```

Comment:
Is the parentheses around `-1` necessary here? You seem to have been able remove redundant parentheses in other cases.